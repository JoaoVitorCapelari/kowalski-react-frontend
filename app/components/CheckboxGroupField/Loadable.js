/**
 *
 * Asynchronously loads the component for CheckboxGroup
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
