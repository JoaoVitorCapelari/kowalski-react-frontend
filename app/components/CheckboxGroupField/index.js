/**
*
* CheckboxGroup
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const CheckboxContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  border-radius: 4px;
  border: 1px solid #aaa;
  padding-left: 10px;
  margin-top: 10px;
`;

// TODO export this component to another file
const DefaultCheckboxGroupItem = ({ input, label }) => (
  <label htmlFor={input.id}>
    <input type="checkbox" {...input} />
    <span>{label}</span>
  </label>
);

DefaultCheckboxGroupItem.propTypes = {
  input: PropTypes.shape({
    id: PropTypes.any.isRequired,
  }).isRequired,
  label: PropTypes.string.isRequired,
};

const createChangeHandler = ({ input, input: { onChange, onBlur }, value }) => (event) => {
  // TODO change the code to handle value as a List from immutable
  const arr = [...input.value];
  if (event.target.checked) {
    arr.push(value);
  } else {
    arr.splice(arr.indexOf(value), 1);
  }
  onBlur(arr);
  return onChange(arr);
};

const CheckboxGroupField = ({ input, meta, options, checkboxItemComponent }) => {
  const { name, onFocus } = input;
  const inputValue = input.value;
  const Component = checkboxItemComponent || DefaultCheckboxGroupItem;
  const checkboxes = options.map((option, index) => {
    const { label, value } = option;
    const key = `checkbox-${index}`;

    return (
      <Component
        input={{
          name: `${name}[${index}]`,
          id: `${name}[${index}]`,
          checked: inputValue.includes(value),
          onChange: createChangeHandler({ input, value }),
          value,
          onFocus,
        }}
        {...{
          meta,
          label,
          option,
          options,
          key,
        }}
      />
    );
  });

  return <CheckboxContainer>{checkboxes}</CheckboxContainer>;
};

CheckboxGroupField.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.node.isRequired,
    value: PropTypes.node.isRequired,
  })).isRequired,
  input: PropTypes.shape({
    name: PropTypes.string,
    onFocus: PropTypes.func,
  }),
  meta: PropTypes.object.isRequired,
  checkboxItemComponent: PropTypes.func,
};

export default CheckboxGroupField;
