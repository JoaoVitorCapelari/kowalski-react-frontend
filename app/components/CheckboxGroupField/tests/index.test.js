import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { reduxForm, Field } from 'redux-form/immutable';
import createHistory from 'history/createMemoryHistory';
import { fromJS } from 'immutable';

import CheckboxGroupField from '../index';
import configureStore from '../../../configureStore';

const initialState = {};
const history = createHistory();
const store = configureStore(initialState, history);

const renderComponent = () => {
  const TestingForm =
    reduxForm({
      form: 'test',
    })(
      () => (
        <form>
          <Field
            name="test"
            id="test"
            label="test"
            component={CheckboxGroupField}
            options={[
              { label: 'A', value: 'A' },
              { label: 'B', value: 'B' },
            ]}
          />
        </form>
      )
    );
  return mount(
    <Provider store={store}>
      <TestingForm />
    </Provider>
  );
};

describe('<CheckboxGroupField />', () => {
  it('render', () => {
    const renderedComponent = renderComponent();
    expect(renderedComponent.html()).toMatchSnapshot();
  });

  it('updates the state accordingly', () => {
    const renderedComponent = renderComponent();
    const getStoreValue = () => (store.getState().get('form').get('test').get('values') || fromJS({})).get('test');
    expect(getStoreValue()).toBeFalsy();
    renderedComponent.find('input[type="checkbox"][value="A"]').first().simulate('change', { target: { checked: true } });
    expect(getStoreValue()).toEqual(fromJS(['A']));
    renderedComponent.find('input[type="checkbox"][value="B"]').first().simulate('change', { target: { checked: true } });
    expect(getStoreValue()).toEqual(fromJS(['A', 'B']));
    renderedComponent.find('input[type="checkbox"][value="B"]').first().simulate('change', { target: { checked: false } });
    expect(getStoreValue()).toEqual(fromJS(['A']));
    renderedComponent.find('input[type="checkbox"][value="A"]').first().simulate('change', { target: { checked: false } });
    expect(getStoreValue()).toEqual(fromJS([]));
  });
});
