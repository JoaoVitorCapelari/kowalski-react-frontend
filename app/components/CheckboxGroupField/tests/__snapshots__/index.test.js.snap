// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxGroupField /> render 1`] = `"<form><div class=\\"CheckboxGroupField__CheckboxContainer-jPIDgp bQvIuF\\"><label for=\\"test[0]\\"><input type=\\"checkbox\\" name=\\"test[0]\\" id=\\"test[0]\\" value=\\"A\\"><span>A</span></label><label for=\\"test[1]\\"><input type=\\"checkbox\\" name=\\"test[1]\\" id=\\"test[1]\\" value=\\"B\\"><span>B</span></label></div></form>"`;
